engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java

```diff
@@ -57,7 +57,15 @@ public final class Module implements EnsoObject {
   private final Map<Source, Module> allSources = new WeakHashMap<>();
   private final Package<TruffleFile> pkg;
   private CompilationStage compilationStage = CompilationStage.INITIAL;
-  private boolean isIndexed = false;
+
+  enum IndexState {
+    NotIndexed,
+    NeedsIndexing,
+    Indexed
+  }
+
+  private IndexState indexState;
```

Comment:
Let me assume the point of this `IndexState` is to have a _tri-state_ status which can be invalidated, scheduled for recomputation and consumed, when it is available. I believe such a concept is best expressed by a [Future](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html). If we change the code in `Module` to just:
```java
private volatile Future<Object> index;

public final vpod invalidateIndex(Future<Object> newIndex) {
  this.index = newIndex; 
}

public final Future<Object> index() {
  return this.index;
}
```
I'll consider it to be _encapsulated enough_ (for now). @hubertp, is such a _future based interface_ enough for your purposes?

Reply:
I don't believe that encapsulated the whole logic as it wouldn't express a situation when in-progress indexing would be marked as dirty and require re-indexing. I can be obviously wrong but there was a reason why logic was added to `Analyze*Job` to deal with this scenario without locking everything, as it was.

Reply:
I continue to believe `Future` is the best concept to represent the desired behavior.

> it wouldn't express a situation when in-progress indexing would be marked as dirty and require re-indexing

No, (cancellable) `Future` represents such concepts neatly, I believe.